### DISCUSSION TITLE
Provide pageContext in prerender() and amend in onBeforeRender()

### DISCUSSION THREAD

**Discussion by lutzissler** [vikejs/vike#722] [Discussion]
I have a Vue app with hundreds of pages generated from the same basic data. To speed up prerendering, I am fetching the data once and then use the `prerender()` hook to provide it in the `pageContext` for all those pages, [as the docs suggest](https://vite-plugin-ssr.com/prerender#for-providing-pagecontext).

I am using Pinia though. With Pinia, I have to get the `initialStoreState` in `onBeforeRender()` and include that in `pageContext` as well, as shown in the [example](https://github.com/brillout/vite-plugin-ssr/blob/main/examples/vue-pinia/renderer/_default.page.server.js#L34). Obviously the state is different for each of the dynamic routes, so I cannot provide `initialStoreState` from the `prerender()` hook.

Unfortunately, because I already return a `pageContext` in my `prerender()` hook, `onBeforeRender()` is not called during prerendering anymore. Ideally, it would still be called so I could *amend* the `pageContext` already provided by `prerender()` and put the `initialStoreState` there.

Is there any way to achieve that?

**Comment by brillout** [comment]
Have you tried to define one gloabl `prerender()` hook in `_default` (doing the "fetching the data once") + multiple `prerender()` hooks on a per-page basis (doing the `pageContext` amending)? VPS will call of them.

**Reply by lutzissler** [reply]
As always, thanks for the quick response! It doesn’t look like this will work because in the per-page `prerender()` hooks I don’t have access to the `pageContext` that I create/populated in the global `prerender()` hook. In other words:

```js
// _default.page.server.js
export function prerender() {
  const data = … // Some data fetching

  // Set some global pageContext as all the pages need the data
  const pageContext = {
    data
  }

  const pages = data.items.map(item => {
    url: `/items/${item.id}`,
    pageContext
  })
}


// items/item.page.server.js:
export function prerender(pageContext) {
  // Would expect to receive pageContext here because I need the ‘data’ property
  // and want to return an amended pageContext
  // But pageContext is undefined here
  …
}
```

**Reply by lutzissler** [reply]
Providing the actual `pageContext` from the global `prerender()` might be the wrong approach in my case. Instead, I would need to provide some ‘seed data’ to the per-page `onBeforeRender()` hook but **outside** `pageContext`. `onBeforeRender()` could populate the store with that ‘see data’ (instead of loading it from the server for every page during prerender), and then get the `initialStoreState` to put into `pageContext`.

Is there any way to achieve that in VPS, i. e. to pass data from `prerender()` to `onBeforeRender()` outside of `pageContext`?

**Reply by brillout** [reply]
My bad, you're right.

Another trick I can think of: populate `prerenderContext.globalData` in your single gloabl `prerender()` hook, and don't use your per-page `prerender()` hooks but your `onBeforeRender()` per-page hooks in order to define the `pageContext` of each URL. (Make sure that your `prerender()` hooks don't define the `pageContext` of URLs so that VPS calls your `onBeforeRender()` hooks.)

**Reply by brillout** [reply]
> populate `prerenderContext.globalData` in your single gloabl `prerender()` hook

I'm not sure if it's possible though, it needs to be tested.

Another alternative: the [`prerender()` API](https://vite-plugin-ssr.com/prerender-programmatic).

**Reply by brillout** [reply]
IIRC in `onBeforeRender()` it's available as something like `pageContext._prerenderContext`.

**Reply by lutzissler** [reply]
> populate `prerenderContext.globalData` in your single gloabl `prerender()` hook

How would I do that? I tried

```js
export function prerender() {
  return {
    url: '…',
    prerenderContext: {
      globalData: { … }
    }
  }
}
```

but that is giving me an error (The prerender() hook defined by /src/renderer/_default.page.server.js returned an invalid value: unexpected object key `prerenderContext`.)

**Reply by lutzissler** [reply]
Looking at the source of VPS that clearly was the wrong way as `prerenderContext` is some global thing ;) Only I don’t see how I would access that in my `prerender()` hook. Any advice would be very much appreciated.

**Reply by brillout** [reply]
Yes that'd need to be implemented but in the meantime you can mutate `prerenderContext`.

```js
export function prerender(prerenderContext) {
  prerenderContext.globalData = ...
}
```

Although I'm not sure if you can access `prerenderContext` in `onBeforeRender()` (that may need to be implemented).

**Reply by brillout** [reply]
> Looking at the source of VPS that clearly was the wrong way as `prerenderContext` is some global thing ;) Only I don’t see how I would access that in my `prerender()` hook. Any advice would be very much appreciated.

You're right for the per-page `prerender()` hooks, but the global `prerender()` does have access to `prerenderContext` (see for example how i18n works for pre-rendered apps). FYI the hooks have been renamed with a much clearer naming in the V1 design.

**Reply by lutzissler** [reply]
> You're right for the per-page `prerender()` hooks, but the global `prerender()` does have access to `prerenderContext`

Not sure if I get that right, but when I do

```js
// _default.page.server.js
export async function prerender(prerenderContext) {
    console.log('prerenderContext', prerenderContext)
}
```

in my console I get an `undefined`.

**Reply by brillout** [reply]
My bad, the global one is actually called `onBeforePrerender()`. (The current naming really is confusing 😅. Especially because `prerender()` hooks are actually called *before* `onBeforePrerender()`. Can't wait to ship the V1 design with improved naming.)

Example: https://vite-plugin-ssr.com/i18n#pre-rendering.

**Reply by lutzissler** [reply]
I’ve tested this, but as you mentioned earlier, I cannot access `prerenderContext` in `onBeforeRender()`. I’ve tried as `pageContext._prerenderContext` and also checked `pageContext` for anything similar, without luck.

**Reply by brillout** [reply]
Let me check.

**Reply by brillout** [reply]
Just released `0.4.99` which adds `pageContext._prerenderContext`.

**Reply by lutzissler** [reply]
Much appreciated! I’ve tested it like this:

```js
// _default.page.server.js
export async function onBeforePrerender(prerenderContext) {
    prerenderContext.globalData = {
        test: 1
    }
    return { prerenderContext }
}

export async function onBeforeRender(pageContext) {
    // This logs as undefined
    console.log(pageContext._prerenderContext.globalData)
    …
}
```

What am I missing?

**Reply by brillout** [reply]
Try `(pageContext._)prerenderContext.pageContexts.gloablData`

**Reply by lutzissler** [reply]
Ok, but even this doesn’t work:

```js
// _default.page.server.js
export async function onBeforeRender(pageContext) {
    console.log(pageContext._prerenderContext)
    …
}
```

Shouldn’t that work after updating to `0.4.99`?

**Reply by brillout** [reply]
It should, see the fix commit. Hm, not sure what's going on.

**Reply by lutzissler** [reply]
My bad, I had made some changes while testing and that left the `prerender()` hook provide `pageContext`. Without that, it’s working (of course).

**Reply by lutzissler** [reply]
The only page for which `pageContext._prerenderContext` is now missing in `onBeforeRender()` is for the error page. I can handle that case of course but it would be great for consistency if `pageContext._prerenderContext` would be present there, too.

**Reply by brillout** [reply]
Yea, in general, the whole DX here would need to be polished (I've some ideas). Feel free to open a new Feature Request describing your use case in generic terms.

Also, in case you think you can convince your company: https://github.com/sponsors/brillout &ndash; sponsorships are making a huge impact.

Curious: how did you stumble upon VPS?
